### ISSUE TITLE
FedEx not updating

### ISSUE THREAD

**Issue by iankaufmann** [moralmunky/Home-Assistant-Mail-And-Packages#221]
I have an incoming FedEx package scheduled for delivery today.  

There is an email notification in my inbox in the specified folder from "TrackingUpdates@fedex.com" with the subject "Your package is scheduled for delivery today" from 7:18am this morning.

The integration last scanned at 9:04am.

It picked up my 3 USPS pieces of mail, so I know that is working, but nothing from FedEx.

In the wiki it says it checks for:
"Your package is scheduled for delivery today" in the subject, which matches.

It also says it checks for:
"Delivery scheduled for today" in the body, but that does not match.

Instead, in the body, it says "Your package from ******* is scheduled for delivery today".

<img width="702" alt="Screen Shot 2020-09-25 at 9 11 14 AM" src="https://user-images.githubusercontent.com/1100001/94278072-0532db00-ff10-11ea-8a5b-eeac4212ce0e.png">

<img width="371" alt="Screen Shot 2020-09-25 at 9 12 48 AM" src="https://user-images.githubusercontent.com/1100001/94278093-0c59e900-ff10-11ea-84c2-f121e474639a.png">

<img width="454" alt="Screen Shot 2020-09-25 at 9 13 32 AM" src="https://user-images.githubusercontent.com/1100001/94278105-10860680-ff10-11ea-909a-c9509e413353.png">

<img width="453" alt="Screen Shot 2020-09-25 at 9 13 42 AM" src="https://user-images.githubusercontent.com/1100001/94278117-14b22400-ff10-11ea-952a-0ee9e5621e0b.png">

Let me know if there is any other info that I should provide.

Thanks!

**Comment by firstof9** [comment]
Could you enable debugging and post the results from the log?

**Comment by iankaufmann** [comment]
Is there a specific option in the integration to enable debugging?  I don't see an option for that in the config panel.

I do see this in the HA log:

```
Logger: homeassistant.helpers.entity
Source: /usr/local/lib/python3.8/asyncio/events.py:81
First occurred: 9:50:37 AM (1 occurrences)
Last logged: 9:50:37 AM
Update of sensor.mail_fedex_delivered is taking over 10 seconds 
```

**Comment by firstof9** [comment]
Yup [here's the link to the wiki](https://github.com/moralmunky/Home-Assistant-Mail-And-Packages/wiki/Troubleshooting#enable-debugging) on how to do it.

**Comment by iankaufmann** [comment]
Sorry I didn't see that, I clicked the wiki and searched for "debug" but didn't even see the troubleshooting page.  Thanks for being cool about it!

Here is the DEBUG log from the last run:
```
2020-09-25 11:04:55 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Config scan interval: 0:02:00
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 1 (Your package has been delivered)
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 1 (Your package has been delivered) results: b'' count: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 1 (Delivery scheduled for today)
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 1 (Delivery scheduled for today) results: b'' count: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 2 (Your package is scheduled for delivery today)
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 2 (Your package is scheduled for delivery today) results: b'' count: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (mcinfo@ups.com) with subject 1 (Your UPS Package was delivered)
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Searching for (Tracking Number) in (0) emails
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (Tracking Number) count results: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (mcinfo@ups.com) with subject 1 (Your UPS Package was delivered) results: b'' count: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (mcinfo@ups.com) with subject 2 (Your UPS Packages were delivered)
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Searching for (Tracking Number) in (0) emails
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (Tracking Number) count results: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (mcinfo@ups.com) with subject 2 (Your UPS Packages were delivered) results: b'' count: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (mcinfo@ups.com) with subject 1 (UPS Update: Package Scheduled for Delivery Today)
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (mcinfo@ups.com) with subject 1 (UPS Update: Package Scheduled for Delivery Today) results: b'' count: 0
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (mcinfo@ups.com) with subject 2 (UPS Update: Follow Your Delivery on a Live Map)
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (mcinfo@ups.com) with subject 2 (UPS Update: Follow Your Delivery on a Live Map) results: b'' count: 0
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (auto-reply@usps.com) with subject 1 (Item Delivered)
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (auto-reply@usps.com) with subject 1 (Item Delivered) results: b'' count: 0
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (auto-reply@usps.com) with subject 1 (Expected Delivery on)
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Searching for (Your item is out for delivery) in (0) emails
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (Your item is out for delivery) count results: 0
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (auto-reply@usps.com) with subject 1 (Expected Delivery on) results: b'' count: 0
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find Informed Delivery mail
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Cleaning up image directory: /config/www/mail_and_packages/
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Informed Delivery email found processing...
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Extracting image from email
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Extracting image from email
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Extracting image from email
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Extracting image from email
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Extracting image from email
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Removing duplicate images.
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Removing USPS announcement images.
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Image Count: 3
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Resizing images to 724x320...
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Creating array of image files...
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Generating animated GIF
2020-09-25 11:04:57 INFO (MainThread) [custom_components.mail_and_packages.sensor] Mail image generated.
2020-09-25 11:04:57 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find Amazon email with item list ...
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon emails found: 1
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Email Multipart: True
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Content Type: multipart/alternative
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon emails found: 0
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon emails found: 0
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon emails found: 0
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon Count: 0
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find Amazon email with item list ...
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon emails found: 1
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Email Multipart: True
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Content Type: multipart/alternative
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon emails found: 0
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon emails found: 0
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon emails found: 0
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon order: ['#112-1279225-4568241']
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Searching for Amazon delivered email(s)...
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon delivered email(s) found: 0
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Searching for Amazon image in emails...
2020-09-25 11:04:58 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] HTML Amazon emails found: 0
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon delivered email(s) found: 0
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Searching for Amazon image in emails...
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] HTML Amazon emails found: 0
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon delivered email(s) found: 0
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Searching for Amazon image in emails...
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] HTML Amazon emails found: 0
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Amazon delivered email(s) found: 0
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Searching for Amazon image in emails...
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] HTML Amazon emails found: 0
2020-09-25 11:04:59 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Updated scan time: Sep-25-2020 11:04 AM
```

**Comment by firstof9** [comment]
Looks like it doesn't like the `" OR FROM "` I added to streamline email queries 😞

**Comment by iankaufmann** [comment]
No worries!  It's always something!

**Comment by firstof9** [comment]
I think is see the issue, stand by

**Comment by firstof9** [comment]
Give b28 a shot, should fix it right up.

**Comment by iankaufmann** [comment]
I updated via HACS and restarted HA.

Unfortunately, all sensors came up as unavailable, with this in the log:

```
2020-09-25 11:55:37 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up mail_and_packages platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 193, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/mail_and_packages/sensor.py", line 74, in async_setup_entry
    sensors.append(PackagesSensor(data, variable, unique_id))
  File "/config/custom_components/mail_and_packages/sensor.py", line 207, in __init__
    self.unique_id = unique_id
AttributeError: can't set attribute
```


Tried restarting HA twice and got the same thing.

**Comment by firstof9** [comment]
Try going thru the `Options` menu and resubmit your info.
![image](https://user-images.githubusercontent.com/1105672/94295987-f4399800-ff16-11ea-9f04-763f4f4747dc.png)

**Comment by iankaufmann** [comment]
Forgot to mention that I did try that too...

I also just now removed the integration completely and re-added it, and I'm seeing the same error in the log.

`Error while setting up mail_and_packages platform for sensor`

**Comment by iankaufmann** [comment]
I wonder if it's because I went from b26 to b28.  I see some changes to entity_id in b27.  

Rolling back to b26 made the sensors come back, so I will try to update to b27, restart, then update to b28...

**Comment by iankaufmann** [comment]
OK so I just updated to b27 and restarted, and I have the same issue.

```
2020-09-25 11:55:37 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up mail_and_packages platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 193, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/mail_and_packages/sensor.py", line 74, in async_setup_entry
    sensors.append(PackagesSensor(data, variable, unique_id))
  File "/config/custom_components/mail_and_packages/sensor.py", line 207, in __init__
    self.unique_id = unique_id
AttributeError: can't set attribute
```

So this seems to be related to b27, not b28.

Hope that helps!

**Comment by firstof9** [comment]
Oh ooops im dumb :P one sec

**Comment by firstof9** [comment]
There b29

**Comment by firstof9** [comment]
Note this will cause your old sensors to need removal and rename the new ones.

**Comment by iankaufmann** [comment]
Yeah I saw that I ended up with a bunch of duplicates with "_2" appended to their names.

I just removed the integration again, and set it up fresh again.

Now I get this:

```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 193, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/mail_and_packages/sensor.py", line 74, in async_setup_entry
    sensors.append(PackagesSensor(data, variable, unique_id))
  File "/config/custom_components/mail_and_packages/sensor.py", line 208, in __init__
    self.update()
  File "/config/custom_components/mail_and_packages/sensor.py", line 262, in update
    self.data.update()
  File "/usr/src/homeassistant/homeassistant/util/__init__.py", line 239, in wrapper
    result = method(*args, **kwargs)
  File "/config/custom_components/mail_and_packages/sensor.py", line 155, in update
    total = data[delivering] + data[delivered]
KeyError: 'dhl_delivering'
```

However, if I uncheck the 3 "DHL" options from the list, everything else works... so that's good.

The last bit of unfortunate news I have for you is that FedEx is still showing zero.

But hey, at least we found the entity_id issue!

**Comment by firstof9** [comment]
is it still reporting:
```
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 1 (Your package has been delivered)
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 1 (Your package has been delivered) results: b'' count: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 1 (Delivery scheduled for today)
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 1 (Delivery scheduled for today) results: b'' count: 0
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Attempting to find mail from (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 2 (Your package is scheduled for delivery today)
2020-09-25 11:04:56 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Search for (TrackingUpdates@fedex.com, fedexcanada@fedex.com) with subject 2 (Your package is scheduled for delivery today) results: b'' count: 0
```
?

**Comment by firstof9** [comment]
Try modifying line 608 to this: `email_search = '" OR FROM "'.join(email)`

**Comment by firstof9** [comment]
Side note: this is why 0.3.x has unit tests 😜

**Comment by iankaufmann** [comment]
Haha that will be nice to have tests... I can also appreciate how hard this is to debug when you don't have a FedEx email sitting in your mailbox to test with.

Here is what I've found now...

Changing to `email_search = '" OR FROM "'.join(email)` had no effect.

After looking at the `const.py` and `sensor.py` files more, it looks like the subject line that matches is actually "subject_2".  The email that I have in my inbox doesn't match "subject"...

So Looking in `sensor.py`, it looks like only the "subject" check was modified to include multiple email addresses in the search.  If you look at the logic for "subject_2", it is only checking for a single email address.

I tried to modify that by copying the logic from the "subject" test down to "subject_2" but it still didn't work.

I still think the logic/syntax for the join() isn't right.

However, I was able to get my sensor to show 1 package by changing `const.py` to:
```
FEDEX_Email = "TrackingUpdates@fedex.com"
FEDEX_Delivering_Subject = "Your package is scheduled for delivery today"
FEDEX_Delivering_Subject_2 = "Your package is scheduled for delivery today"
```

Basically, I just copied subject 2 into subject 1, and made it only check the one email address (I'm not in Canada so I don't need both anyway).

Maybe that will point you in the right direction?

Thanks for your help so far and for inspiring me to take a look under the hood.

**Comment by firstof9** [comment]
Ya something is going on with the join it seems, I'll do some more testing and update this issue.

**Comment by firstof9** [comment]
Looks like the `const.py` just needs updating:
```
FEDEX_Email = ["TrackingUpdates@fedex.com, fedexcanada@fedex.com"]
```

**Comment by firstof9** [comment]
Let me know how this one goes, should fix you right up.

**Comment by ronaldheft** [comment]
FedEx is working with b30!

**Comment by firstof9** [comment]
If this is also fixed for you @iankaufmann go ahead and close the issue.

**Comment by iankaufmann** [comment]
Happy to report it's working great!

Thanks a ton!
